napari/layers/utils/style_encoding.py

```diff
@@ -0,0 +1,228 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterable, Optional, Sequence, Tuple, Union
+
+import numpy as np
+from pydantic import ValidationError, parse_obj_as, validator
+
+from ...utils import Colormap
+from ...utils.colormaps import ValidColormapArg, ensure_colormap
+from ...utils.colormaps.categorical_colormap import CategoricalColormap
+from ...utils.colormaps.standardize_color import transform_color
+from ...utils.events import EventedModel
+
+
+class StyleEncoding(EventedModel, ABC):
+    array: np.ndarray = []
+
+    @abstractmethod
+    def _apply(
+        self, properties: Dict[str, np.ndarray], indices: Sequence[int]
+    ) -> np.ndarray:
+        pass
+
+    def update_all(self, properties: Dict[str, np.ndarray], n_rows: int):
+        indices = range(0, n_rows)
+        self.array = self._apply(properties, indices)
+
+    def update_tail(self, properties: Dict[str, np.ndarray], n_rows: int):
```

Comment:
this should update the last `n_rows` right? I'm a little confused about the `update_all` and `update_tail`? Should it be `update_head` and `update_tail`? seems more symmetric

Reply:
See comment above, as `n_rows` means the same thing here - the total number of rows in the `properties` table.

Maybe I should rename it to `total_rows`? Or happy to consider other method names.

One thing to keep in mind is that I think this signature might change, especially if we introduce some kind of `PropertyTable` class (even a wrapper) that would have this state.

Reply:
Documenting these methods seems like a good idea.

Reply:
Done - I tried to add docstrings on all new public classes, attributes, functions. Hope that helps.